aries_vcx/src/handlers/connection/inviter/state_machine.rs

```diff
@@ -459,17 +458,38 @@ impl SmConnectionInviter {
     }
 
     pub fn handle_ack(self, ack: Ack) -> VcxResult<Self> {
-        let Self { source_id, pairwise_info, state, send_message } = self;
+        let Self { source_id, pairwise_info, state, send_message } = self.clone();
         let state = match state {
             InviterFullState::Responded(state) => {
-                InviterFullState::Completed((state, ack).into())
+                let thread_id = self.get_thread_id()?;
```

Comment:
Can't we do this in `update_state_with_message`? So we wouldn't have to this explicitly on every particular message handler?

Reply:
In principle, we can do what you suggest, but we don't need the thread in every handler or state, so to me it makes sense to fetch it on demand (it returns result, so this way we don't return error when `get_thread_id()` fails for some reason yet we don't need to use thread id in the state). It also reduced the number of arguments to some of the handlers. But you are right it would maybe remove some lines.

Reply:
Well right now we pretty much only check thread id in received message for `ack` message. Why? Whatever message we try to update the state machine with, it should have expected thread id. You should either add these checks to all other handles, or pull the checking on `update_state_with_message` (which I strongly favor). 

Equally there's forgotten cases to check for thread id on invitee side. For example we check `ack`'s thread_id in inviter side, but not on invitee's side.

Reply:
As per worries about calling `get_thread_id` and getting error, it should generally not happen if received messages are properly validated - this relates to my comments here https://github.com/hyperledger/aries-vcx/pull/348/files#r714713662